README.md
# AngularTS WP

> A starting point for building web applications with AngularJS 1.6 using TypeScript, Webpack and PostCSS.


## Prerequisites

- [Node.js](https://nodejs.org/en/download/).
- [npm](https://www.npmjs.com/), installed with Node.js.
- [yarn](https://yarnpkg.com/) a package manager for project dependencies.


## Install dependencies

#### 1) Check `Node.js` version.

```sh
node --version
```
The version should be at or above 6.9

If you don't have Node.js installed go to [nodejs](https://nodejs.org/en/download/) and install the appropiate version or use [nvm](http://www.sergiolepore.net/2014/06/30/nvm-instalando-y-usando-node-version-manager/) (Recommended).

#### 2) Check `yarn` version.

```sh
yarn --version
```
The version should be at or above 0.20.0

If you don't have yarn installed go to [yarn](https://yarnpkg.com/en/docs/install) and install the appropiate version.


## Workflow

### Environment variables

```sh
##
# angularts-wp environment variables
##

# Host configuration
## Environment for server.
export NODE_ENV="development"
## Host of webapplication.
export ANGULARTS_WP_HOST=localIP
## Port of webapplication (Use in development).
export ANGULARTS_WP_PORT=3000

# Access URLs
## Url of API.
export ANGULARTS_WP_API_URL="http://${ip.address()}:${8080}/api"
```

You can copy this script in bashrc file, modify the variables and delete variables that don't need modification or are undefined.

To change any of this variables:

```sh
export ANGULARTS_WP_PORT=8000
```

### Development workflow

#### Install the workspace dependencies.

```sh
# cd to project folder
yarn
```

#### Static server with live reload
Create a server using `webpack-dev-server` to serve the application in development environment.
The browser reloads the app when any file change:

```sh
# cd to project folder
yarn start
# After this, a message indicate the url to run the application
```

### Production workflow

You can use two ways:

#### The repository is in `production` server
- Set the environment variables with production values. `export NODE_ENV='production'`, ...
- Run:

```sh
# cd to project folder
# Install the workspace dependencies.
# This npm script force the installation of development dependencies.
yarn install:dev

# Build
yarn build
```
After run this script the `dist` folder is ready for production.
> Run the previous script again every time that update the repository.

#### Build in `development` environment and upload the files to production server.
- Verify that all dependencies are installed.
- Set the environment variable with production values
- Build the application

```sh
# cd to project folder
yarn build
```
After this, all files are ready in `dist` folder to upload in any production server.

:warning: Reset the environment variables to `development` values.

### Install/Uninstall/Update dependencies
Follow this rules to update dependencies:

- Install dependencies

```sh
# To install production dependencies
# cd to project folder
yarn add angular --exact

# To install development dependencies
# cd to project folder
yarn add eslint --exact --dev
```

- Uninstall dependencies

```sh
# cd to project folder
yarn remove angular
```

- **To Upgrade any dependencies follow `Uninstall dependencies` step and install again like `Install dependencies`.**

### Others scripts

To run eslint in console type:

```sh
# cd to project folder
yarn lint
```


## Style Guides

### TypeScript

**[Nimedev TypeScript style guide](https://github.com/nimedev/typescript)**.

### JavaScript

**[Nimedev JavaScript style guide](https://github.com/nimedev/javascript)**.

### CSS

**[niduscss-framework](https://github.com/nimedev/niduscss-framework)**.

**[SuitCSS naming conventions](https://github.com/suitcss/suit/blob/master/doc/naming-conventions.md)**.


## Angular HTML5 mode or pretty URLs

This mode is set by default. If use it consider this:

- All call related to links like `href="/link/to/anyware"`, `$location.path('/link/to');`, etc. result as a hit to server. Use it only for SEO porpuses.
- To avoid unnecesary hits to server (SPA) you can use `ui-sref="/link/to"` in html files or `$state` service in angular files.


## Editors
This project is configured to use with `VScode`. Also contain pre-configured task for development. See `.vscode` folder to explore the options.


## [Contributing](CONTRIBUTING.md)


## [Changelog](CHANGELOG.md)


## [License](LICENSE.md)


CHANGELOG.md
# Change Log

## [6.2.2] - 2017-03-31

### Updated
- Upgrade dep: angular.
- Upgrade devDep: @types/angular and eslint.


## [6.2.1] - 2017-03-29

### Updated
- Upgrade devDep: @types/angular, autoprefixer, babel-core, babel-eslint, babel-preset-es2015, eslint, typescript, webpack, webpack-dev-server and webpack-merge.


## [6.2.0] - 2017-03-09

### Modified
- Use webpack-kit-nimedev-ts to config webpack.


## [6.1.1] - 2017-03-09

### Fixed
- Update package version.


## [6.1.0] - 2017-03-09

### Modified
- Use npm @types/angular instead angular and jquery typings.


## [6.0.0] - 2017-03-09

### Modified
- Improve colors palette.

### Deleted
- Remove niduscss utils packages.


## [5.7.0] - 2017-03-09

### Added
- Use webpack-parts-nimedev to config webpack.

### Updated
- Upgrade dep: angular.
- Uprade devDep: autoprefixer, eslint, ip, tslint and webpack-merge.


## [5.6.0] - 2017-02-27

### Updated
- Upgrade devDep: autoprefixer, extract-text-webpack-plugin, file-loader, html-loader, eslint-loader, postcss-loader, style-loader, stylelint, ts-loader, tslint-loader, eslint, typescript, css-loader, url-loader, babel-loader, webpack-dev-server, babel-core, webpack-merge and opener.


## [5.5.0] - 2017-02-10

### Deleted
- Remove niduscss-lib-mixins dependency.

### Updated
- Upgrade dep: angular, niduscss-utils-align, niduscss-utils-shadow and niduscss-utils-text.
- Upgrade devDep: eslint and typescript.


## [5.4.0] - 2017-02-02

### Deleted
- Remove .jsbeautifyrc file.

### Updated
- Upgrade devDep: postcss-color-function, postcss-custom-properties, stylelint and webpack-dev-server


## [5.3.0] - 2017-01-31

### Modified
- Improve README.md documentation.

### Updated
- Upgrade devDep: autoprefixer, eslint, file-loader, html-webpack-plugin, tslint, webpack and webpack-merge.


## [5.2.0] - 2017-01-24

### Modified
- Improve js-beautify settings.
- Improve dev-server configuration for webpack.


## [5.1.1] - 2017-01-22

### Fixed
- Fix problem loading images in css files.


## [5.1.0] - 2017-01-22

### Updated
- Upgrade devDep: autoprefixer.


## [5.0.0] - 2017-01-22

### Modified
- Migrate to webpack 2.


## [4.8.0] - 2017-01-19

### Added
- Add babel support.


## [4.7.0] - 2017-01-19

### Updated
- Upgrade devDep: eslint-config-nimedev-base.


## [4.6.3] - 2017-01-14

### Fixed
- Enable sourcemaps in tsconfig file.


## [4.6.2] - 2017-01-14

### Fixed
- Fix documentation issues.


## [4.6.1] - 2017-01-14

### Fixed
- Fix issues in workflow documentation.


## [4.6.0] - 2017-01-14

### Added
- Add install:dev scripts.

### Modified
- Improve dev-server module in webpack.parts lib.
- Updated workflow documentation.


## [4.5.0] - 2017-01-12

### Added
- Add style guide section to README.md file.

### Updated
- Upgrade eslint-config-nimedev-base, postcss-loader, typescript and webpack-merge.


## [4.4.0] - 2017-01-10

### Modified
- Improve documentation for environment variables.

### Updated
- Upgrade clean-webpack-plugin, eslint, postcss-import, tslint and webpack-merge.


## [4.3.0] - 2017-01-05

### Updated
- Upgrade autoprefixer, eslint-config-nimedev-base, html-webpack-plugin, stylelint and webpack-merge.


## [4.2.0] - 2016-12-29

### Added
- Add tslint-config-nimedev rules.

### Modified
- Change angular version in docs to 1.6.


## [4.1.0] - 2016-12-28

### Updated
- Improve vscode settings.


## [4.0.1] - 2016-12-28

### Added
- Add .nvmrc file.


## [4.0.0] - 2016-12-28

### Added
- Add .jsbeautifyrc file.
- Add .editorconfig file.

### Modified
- Update tslint rules.
- Implement eslint-config-nimedev-base.


## [3.4.0] - 2016-12-27

### Modified
- Update tsconfig.json and typings.json.

### Updated
- Upgrade angular, eslint, postcss-loader, postcss-mixins, pre-commit, autoprefixer, ts-loader, tslint, typings, webpack-merge, webpack-validator and stylelint.


## [3.3.1] - 2016-12-12

### Updated
- Use node v6.9.2.
- Upgrade angular, eslint, ts-loader, typescript, webpack, webpack-merge and webpack-validator.


## [3.3.0] - 2016-12-05

### Updated
- Upgrade angular, css-loader, eslint, postcss-import, postcss-loader, tslint, tslint-loader, webpack-merge and webpack-validator.


## [3.2.1] - 2016-11-22

### Modified
- Change yarn version and remove typings step in README.md file.


## [3.2.0] - 2016-11-22

### Updated
- Upgrade tslint and tslint-loader.


## [3.1.4] - 2016-11-21

### Deleted
- Remove noUnusedParameters from tsconfig.json.

### Modified
- Add IComponentOptions interface to component object in app.component.ts.


## [3.1.3] - 2016-11-20

### Fixed
- Move tslint to devDependencies.


## [3.1.2] - 2016-11-20

### Updated
- Disable javascript validation in vscode settings.


## [3.1.1] - 2016-11-20

### Fixed
- Change year in CHANGELOG file.


## [3.1.0] - 2016-11-20

### Modified
- Update tsconfig and tslint configuration files.
- Update vscode settings.

### Updated
- Upgrade stylelint and tslint-loader.


## [3.0.0] - 2016-11-17

### Modified
- Change folder structure.
- Use environment variables to set app constants.

### Updated
- Upgrade css-loader.


## [2.4.0] - 2016-11-16

### Added
- Minify index.html when build.

### Modified
- Improve vscode settings.

### Updated
- Upgrade autoprefixer, copy-webpack-plugin, eslint, postcss-import, postcss-loader, ts-loader, typescript and webpack-merge.


## [2.3.0] - 2016-11-09

### Modified
- Update lint:js script.

### Updated
- Upgrade autoprefixer, clean-webpack-plugin, eslint, html-webpack-plugin, postcss-loader, ip and postcss-import.
- Upgrade ts-loader, typescript and typings.


## [2.2.0] - 2016-10-27

### Modified
- Move some assets to public folder.
- Move opener to webpack config file.

### Updated
- Use nodejs v6.9.1.
- Upgrade niduscss-elements-form and niduscss-lib-mixins.
- Upgrade stylelint, ts-loader, copy-webpack-plugin, html-webpack-plugin and webpack.


## [2.1.1] - 2016-10-18

### Updated
- Upgrade eslint, ts-loader and webpack-merge.


## [2.1.0] - 2016-10-17

### Modified

webpack.config.js
// Set default node environment to development
process.env.NODE_ENV = process.env.NODE_ENV || 'development'

const path = require('path')
const ip = require('ip')
const opener = require('opener')
const stylelint = require('stylelint')
const webpack = require('webpack')
const merge = require('webpack-merge')
const webpackKit = require('webpack-kit-nimedev-ts')

const host = process.env.ANGULARTS_WP_HOST || ip.address()
const port = process.env.ANGULARTS_WP_PORT || 3000
const apiUrl = process.env.ANGULARTS_WP_API_URL || `http://${ip.address()}:${8080}/api`
const PATHS = {
  app: path.join(__dirname, 'app'),
  dist: path.join(__dirname, 'dist'),
  images: path.join(__dirname, 'app/assets/images'),
  icons: path.join(__dirname, 'app/assets/icons'),
  fonts: path.join(__dirname, 'app/assets/fonts')
}

const common = merge([
  // Common settings
  {
    entry: {
      app: PATHS.app
    },
    output: {
      path: PATHS.dist,
      filename: '[name].js'
    },
    plugins: [
      new webpack.DefinePlugin({
        'process.env': {
          NODE_ENV: JSON.stringify(process.env.NODE_ENV),
          ANGULARTS_WP_API_URL: JSON.stringify(apiUrl)
        }
      })
    ],
    resolve: {
      extensions: ['.ts', '.js', '.json', '.css']
    }
  },
  webpackKit.htmlPlugin(),
  webpackKit.lintCSS(stylelint, { include: PATHS.app }),
  webpackKit.loadHtml({ include: PATHS.app }),
  webpackKit.loadTS({ include: PATHS.app }),
  webpackKit.loadImages({
    include: PATHS.images,
    options: {
      name: './assets/images/[name].[hash].[ext]',
      limit: 25000
    }
  }),
  webpackKit.loadSvgSprite({
    include: PATHS.icons,
    options: {
      name: './assets/icons/[name].[hash].[ext]'
    }
  }),
  webpackKit.loadFonts({ include: PATHS.fonts }),
  webpackKit.loadAssets({ include: PATHS.app })
])

module.exports = ({ target }) => {
  // Return production configuration
  if (target === 'production') {
    return merge([
      common,
      {
        output: {
          filename: '[name].[chunkhash].js'
        },
        plugins: [
          new webpack.HashedModuleIdsPlugin()
        ]
      },
      webpackKit.extractBundles(webpack),
      webpackKit.cleanPlugin(PATHS.dist),
      webpackKit.loadJS({ include: PATHS.app }),
      webpackKit.minify(webpack),
      webpackKit.extractCSS({ include: PATHS.app })
    ])
  }

  // Run opener
  opener(`http://${host}:${port}`)

  // Return development configurations
  return merge([
    common,
    {
      plugins: [
        new webpack.NamedModulesPlugin()
      ]
    },
    webpackKit.generateSourcemaps('#inline-source-map'),
    webpackKit.loadCSS({ include: PATHS.app }),
    webpackKit.devServer(webpack, { host, port }),
    webpackKit.loadJS({
      include: PATHS.app,
      eslintOptions: {
        // Emit warnings over errors to avoid crashing
        // HMR on error.
        emitWarning: true
      }
    })
  ])
}
